tempesta_fw/cache.c

```diff
@@ -1077,15 +1075,13 @@ tfw_cache_build_resp_body(TDB *db, TfwHttpResp *resp, TdbVRec *trec,
 					   off, f_size);
 			skb_frag_ref(it->skb, it->frag);
 			ss_skb_adjust_data_len(it->skb, f_size);
-		} else {
-			p = NULL;
-		}
-		if (__tfw_http_msg_add_str_data((TfwHttpMsg *)resp,
-						&resp->body, p, f_size,
-						it->skb))
-			return - ENOMEM;
 
-		++it->frag;
+			if (__tfw_http_msg_add_str_data((TfwHttpMsg *)resp,
+							&resp->body, p, f_size,
+							it->skb))
+				return - ENOMEM;
+			++it->frag;
+		}
```

Comment:
Good catch! `__tfw_http_msg_add_str_data()` call has no sense if we already have something in `resp->body`, but it can lead to serious crash if we just have empty `resp->body` and it's possible if response headers are splitted from the body exactly by TDB page bound.